src/main/java/fact/utils/SamplePedestalEvent.java

```diff
@@ -0,0 +1,350 @@
+    private SourceURL noiseDatabase;
+
+    @Parameter(required = true, description = "The folder containing the data files.")
+    private String dataFolder;
+
+    @Parameter(required = false, defaultValue = "", description = "The value to prepend all keys from the noise file." +
+                    " (Exp: prependKey='Noise_', 'ZdPointing'->'Noise_ZdPointing'")
+    private String prependKey = "";
+
+    @Parameter(required = false, description = "The condition for the noise which will be used")
+    private String noiseCondition ="";
+    private Condition condition;
+
+    @Parameter(required = true, description="The binning of the zenith-angle. If single value: steplenght, bins into the given steplength"+
+                                            "If multivalue: Each value is the length of a single step starting with zenith=0.")
+    private String[] binning;
+
+    @Parameter(required = true, description = "The binning key of the double value from the noise database.")
+    private String dbBinningKey;
+
+    @Parameter(required = true, description = "The binning key of the double value from the item.")
+    private String itemBinningKey;
+
+    // Data for faster binning
+    List<Data> database; // The database containing index,NIGHT,RUN_ID,${dbBinningKey},currents
+    double[] bins;     // The bins containing the max_values of the different bins
+    List<List<Integer>> bins_index;// The indexes of the database that are in each bin
+    Random randGenerator;
+
+    @Override
+    public Data process(Data input) {
+        Utils.isKeyValid(input, this.itemBinningKey, Double.class);
+        double zd_input =(double)input.get(this.itemBinningKey);
+
+        int binNum = this.getBin(zd_input);
+        int sampleSize = this.bins_index.get(binNum).size();
+        if (sampleSize==0) {
+            log.error("Samplesize of bin: "+binNum+" is too small");
+            return null;
+        }
+        int rand = this.randGenerator.nextInt(sampleSize);
+        int index = this.bins_index.get(binNum).get(rand);
+        Data item = null;
+        try {
+            item = this.getNoiseEvent(index);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+        // insert the result into the input item
+        for (String s : item.keySet()) {
+            input.put(this.prependKey+s, item.get(s));
+        }
+
+        return input;
+    }
+
+
+    public void setNoiseCondition(String noiseCondition) {
+        this.noiseCondition = noiseCondition;
+        ConditionFactory factory = new ConditionFactory();
+        this.condition = factory.create(noiseCondition);
+    }
+
+
+    public void setBinning(String[] binning) {
+        this.binning = binning;
+        if (binning.length==1) { // single value
+            double steplength = Double.parseDouble(this.binning[0]);
+            int numBins = (int)ceil(90.0/steplength);
```

Comment:
could this whole code in the setter be encapsulated in a meaningful function and have a unit test?